hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieMergeHandle.java

```diff
@@ -239,7 +240,14 @@ protected void init(String fileId, Iterator<HoodieRecord<T>> newRecordsItr) {
         record.seal();
       }
       // NOTE: Once Records are added to map (spillable-map), DO NOT change it as they won't persist
-      keyToNewRecords.put(record.getRecordKey(), record);
+      String key = record.getRecordKey();
+      if (keyToNewRecords.containsKey(key)) {
```

Comment:
typically we won't hit this, since the precombine is done on incoming records by default, even before we index. So mostly seems like a fallback handling?

Reply:
Yeah, we can avoid this because the deduplication can be executed before entering this handle, but `precombine` is not by default true, take a look at Spark option `DataSourceOptions.INSERT_DROP_DUPS`